### PR TITLE
[api] Overhaul REST API error codes

### PR DESCRIPTION
**Pull request by gregnazario**
### Description
Now error codes should be like 

```
{
    message: String,
    error_code: u64,
    vm_status_code: Option<u64>
}
```

Things to still do:
1. Add ledger info headers to errors
2. Add get table item to BCS
3. Improve simulation error handling (it basically doesn't say anything about the execution status, but it's a different code than the VM status code)

### Test Plan
Let's see if CI still works.

Likely will need to rework the REST API client to deal with the new API codes, and some of the upstream pieces.

Will run a local node to test out as well

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3251)

### CODE REVIEW

**Comment by wrwg** [comment]
Does this [integrate with the new `AbortInfo`](https://github.com/aptos-labs/aptos-core/blob/afe23cdbf7e9f84494d9fe8169d93f1658dd89c9/types/src/transaction/mod.rs#L692)? Or is the string just whatever the error explain gives you? Because this is rather new, just want to be sure you have seen this.

File: api/src/response.rs [hunk 1]
```diff
@@ -379,8 +373,136 @@ pub type BasicResultWith404<T> = poem::Result<BasicResponse<T>, BasicErrorWith40
...
+    build_not_found(
+        "Ledger version",
+        format!("Ledger version({})", ledger_version),
+        AptosErrorCode::VersionNotFound,
+    )
+}
+
+pub fn transaction_not_found_by_version<E: NotFoundError>(ledger_version: u64) -> E {
+    build_not_found(
+        "Transaction",
```
Comment (davidiw) [inline api/src/response.rs:394]: const would be nice

File: api/src/blocks.rs [hunk 2]
```diff
@@ -89,11 +90,44 @@ impl BlocksApi {
...
+                    )
+                } else {
+                    None
+                };
+                let block = Block {
+                    block_height: bcs_block.block_height.into(),
+                    block_hash: bcs_block.block_hash.into(),
+                    block_timestamp: bcs_block.block_timestamp.into(),
+                    first_version: bcs_block.first_version.into(),
+                    last_version: bcs_block.last_version.into(),
```
Comment (davidiw) [inline api/src/blocks.rs:121]: do we actually need this field? last_version = first_version + len of transactions, no?

**Review by davidiw — APPROVED** [review]
please make sure the sdk tests pass and let's get them land blocking again

File: api/src/blocks.rs (hunk 2 above)
Comment (gregnazario) [reply api/src/blocks.rs:121]: You can request a block without the transactions for indexing purposes, then you need the last version.

Comment (bowenyang007) [reply api/src/blocks.rs:121]: How do you get the len of transactions @davidiw? I had to infer len transactions from last and first version b/c that's all you get from the db call.

**Pull request merged**